ballerina/stream_types.bal

```diff
@@ -0,0 +1,80 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/persist;
+
+public class PersistGoogleSheetsStream {
+
+    private stream<record {}, error?>? anydataStream;
+    private persist:Error? err;
+    private string[] fields;
+    private string[] include;
+    private typedesc<record {}>[] typeDescriptions;
+    private GoogleSheetsClient? persistClient;
+    private typedesc<record {}> targetType;
+
+    public isolated function init(stream<record {}, error?>? anydataStream, typedesc<record {}> targetType, string[] fields, string[] include, any[] typeDescriptions, GoogleSheetsClient persistClient, persist:Error? err = ()) {
+        self.anydataStream = anydataStream;
+        self.fields = fields;
+        self.include = include;
+        self.targetType = targetType;
+
+        typedesc<record {}>[] typeDescriptionsArray = [];
+        foreach any typeDescription in typeDescriptions {
+            typeDescriptionsArray.push(<typedesc<record {}>>typeDescription);
+        }
+        self.typeDescriptions = typeDescriptionsArray;
+
+        self.persistClient = persistClient;
+        self.err = err;
+    }
+
+    public isolated function next() returns record {|record {} value;|}|persist:Error? {
+        if self.err is persist:Error {
+            return <persist:Error>self.err;
+        } else if self.anydataStream is stream<record {}, persist:Error?> {
+            var anydataStream = <stream<record {}, error?>>self.anydataStream;
+            var streamValue = anydataStream.next();
+            if streamValue is () {
+                return streamValue;
+            } else if (streamValue is error) {
+                return <persist:Error>error(streamValue.message());
+            } else {
+                record {}|error value = streamValue.value;
+                if value is error {
+                    return <persist:Error>error(value.message());
+                }
+                check (<GoogleSheetsClient>self.persistClient).getManyRelations(value, self.fields, self.include, self.typeDescriptions);
+
+                string[] keyFields = (<GoogleSheetsClient>self.persistClient).getKeyFields();
+                foreach string keyField in keyFields {
+                    if self.fields.indexOf(keyField) is () {
+                        _ = value.remove(keyField);
+                    }
+                }
+                record {|record {} value;|} nextRecord = {value: checkpanic value.cloneWithType(self.targetType)};
+                return nextRecord;
+            }
+        } else {
+            return ();
+        }
+    }
+
+    public isolated function close() returns persist:Error? {
+        check closeEntityStream(self.anydataStream);
```

Comment:
We can move the code out of utils to here since this is the only place where the `closeEntityStream()` method is used

Reply:
Since this method would be simmilar in both in memory and googlesheets, we can reuse it from persist right?

Reply:
Updated in [4154bd8](https://github.com/ballerina-platform/module-ballerinax-persist.googlesheets/pull/1/commits/4154bd8d3c4e336f435e1143a2da09e7129c858c)